File: libsignal-service/src/main/java/org/whispersystems/signalservice/internal/push/RedeemDonationReceiptRequest.java
package org.whispersystems.signalservice.internal.push;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonProperty;

import org.signal.libsignal.zkgroup.receipts.ReceiptCredentialPresentation;

/**
 * POST /v1/donation/redeem-receipt
 *
 * Request object for redeeming a receipt from a donation transaction.
 */
class RedeemDonationReceiptRequest {

  private final String  receiptCredentialPresentation;
  private final boolean visible;
  private final boolean primary;

  /**
   * @param receiptCredentialPresentation base64-encoded no-newlines standard-character-set with-padding of the bytes of a {@link ReceiptCredentialPresentation} object
   * @param visible boolean indicating if the new badge should be visible or not on the profile
   * @param primary boolean indicating if the new badge should be primary or not on the profile; is always treated as false if `visible` is false
   */
  @JsonCreator RedeemDonationReceiptRequest(
      @JsonProperty("receiptCredentialPresentation") String receiptCredentialPresentation,
      @JsonProperty("visible") boolean visible,
      @JsonProperty("primary") boolean primary) {

    this.receiptCredentialPresentation = receiptCredentialPresentation;
    this.visible = visible;
    this.primary = primary;
  }

  public String getReceiptCredentialPresentation() {
    return receiptCredentialPresentation;
  }

  public boolean isVisible() {
    return visible;
  }

  public boolean isPrimary() {
    return primary;
  }
}

File: lib/libsignal-service/src/main/java/org/whispersystems/signalservice/api/donations/RedeemDonationReceiptRequest.kt
/*
 * Copyright 2025 Signal Messenger, LLC
 * SPDX-License-Identifier: AGPL-3.0-only
 */

package org.whispersystems.signalservice.api.donations

/**
 * POST /v1/donation/redeem-receipt
 *
 * Request object for redeeming a receipt from a donation transaction.
 *
 * @param receiptCredentialPresentation base64-encoded no-newlines standard-character-set with-padding of the bytes of a `ReceiptCredentialPresentation`
 * @param visible Whether the new badge should be visible on the profile
 * @param primary Whether the new badge should be primary on the profile; always treated as false if [visible] is false
 */
internal class RedeemDonationReceiptRequest(
  val receiptCredentialPresentation: String,
  val visible: Boolean,
  val primary: Boolean
)
